src/app/content/utils/seoUtils.ts

```diff
@@ -9,20 +12,149 @@ import {
+    '[data-type="abstract"], .learning-objectives, .chapter-objectives, .be-prepared, .os-teacher'
+    ) || [];
 
+  for (let i = 0; i <= introContentList.length; i++) {
+    if (introContentList[i]) {
+      const parent = introContentList[i].parentNode;
+      if (parent === null) {
+        break;
+      }
+      parent.removeChild(introContentList[i]);
+    }
+  }
+};
+
+const getFirstParagraph = (node: HTMLElement) => {
+  if (!node) {
+    return '';
+  }
+  // First look for first p in a section.
+  const firstSection = node.querySelector('section');
+  const first = firstSection
+    ? firstSection.querySelector('p')
+    : node.querySelector('p');
+
+  if (!first || !first.textContent) {
+    return '';
+  }
+
+  // Find first p longer than 100 chars.
+  if (first.textContent.length < 90) {
+    let next = first.nextElementSibling;
+
+    while (next && next.textContent) {
+      if (next.matches('p') && next.textContent.length >= 90) {
+        return next;
+      }
+      next = next.nextElementSibling;
+    }
+  } else {
+    return first;
+  }
+};
+
+export const generateExcerpt = (str: string) => {
+  return str.replace(/\n/g, ' ').replace(/\s\s/g, ' ').trim().substring(0, 152) + '...';
+};
+
+export const createDescription = (loader: AppServices['archiveLoader'], book: Book, page: Page) => {
+  const cleanContent = getCleanContent(book, page, loader);
+  const doc = domParser.parseFromString(cleanContent, 'text/html');
+  const contentNode = doc.body.children[0];
+  const pageType = getPageType(contentNode);
+  const node = assertDefined(
+    findArchiveTreeNodeById(book.tree, page.id),
+    `couldn't find node for a page id: ${page.id}`
+  );
+  const parentPrefix = getParentPrefix(node, true).replace('Ch.', 'Chapter').trim();
+  const parentIsChapter = node.parent ? archiveTreeSectionIsChapter(node.parent) : true;
+  const parentIsBook = node.parent ? archiveTreeSectionIsBook(node.parent) : false;
+  const sectionTitle = getArchiveTreeSectionTitle(node);
+  const parentTitle = node.parent ? getTextContent(node.parent.title) : '';
+
+  if (pageType === 'page') {
+    removeIntroContent(contentNode);
+    const firstP = getFirstParagraph(contentNode);
+    const mathless = firstP ? hideMath(firstP) : null;
+    return mathless
+      ? generateExcerpt(mathless)
+      // tslint:disable-next-line:max-line-length
+      : `On this page you will discover the ${sectionTitle} for ${parentPrefix} of OpenStax's ${book.title} free textbook.`;
+  } else {
+    let descriptionPhrase = '';
+    if (pageType === 'answer-key') {
+      descriptionPhrase = `the Answer Key for ${sectionTitle} of`;
+    } else if (!parentIsChapter && !parentIsBook) {
+      descriptionPhrase = `${parentTitle}: ${sectionTitle} for ${parentPrefix} of`;
+    } else if (sectionTitle !== parentPrefix) {
+      descriptionPhrase = `the ${sectionTitle} for ${parentPrefix} of`;
+    } else {
+      descriptionPhrase = `the ${sectionTitle} for`;
```

Comment:
we're gonna have to integrate this into the internationalization system, so for now what i'd like to see is helpers like this:

```ts
type PageTypes = 'answer-key' | 'section' | 'introduction'; // or whatever, i'm making these up
interface DescriptionTemplateValues { // or whatever, i'm making these up
  parentType: 'chapter' | 'book';
  pageTitle: string;
  bookTitle: string;
  chapterTitle?: string;
}
const generateDescriptionFromTemplate = (pageType: PageTypes, values: DescriptionTemplateValues) {
  switch (pageType) {
    case 'answer-key':
      return `the whole template string`;
    case 'section':
      return `the whole template string`;
    default:
       throw new Error('unknown page type');
  }
}
const getPageType = (treeNode: ArchiveTreeSection, contentNode: HTMLElement): PageTypes => {
  if (whatever) {
    return something;
  }

  return stuff
}
const getPageDescriptionValues = (treeNode: ArchiveTreeSection, contentNode: HTMLElement): DescriptionTemplateValues => {
  return ...;
}
const getPageDescriptionFromContent = (pageType: PageTypes, contentNode: HTMLElement): string | null => {
  // try to find paragraph, if found sanitize and return snippet. maybe use pagetype to filter certain types of content per page, maybe not necessary idk

 // if no paragraph found or content is too short return null
}
const getPageDescription = (... whatever arguments) => {
  const pageType = getPageType(stuff)
  if (['section', 'introduction'].contains(pageType) {

  }
  const contentDescription: string | null = ['section', 'introduction'].contains(pageType)
    ? getPageDescriptionFromContent(stuff)
    : null;

  return contentDescription || generateDescriptionFromTemplate(stuff)
}
```

take all that with a grain of salt i'm trying to outline a sort of logical structure but don't follow through on anything i did if it feels weird or bad when you get into it. do you understand kind of where i'm going with this? eventually we'll replace a chunk of `generateDescriptionFromTemplate` with 
```
intl.formatMessage(`intl:meta-description:${pageType}`, descriptionValues)
```
but i don't want to think about how to get the intl object in there quite yet

Reply:
yep, sounds good

Reply:
still need to do some testing but I thing it's much more legible now. thanks for the outline.